python/tvm/relay/memory_alloc.py

```diff
@@ -0,0 +1,254 @@
+            new_fields.append(field)
+        return expr.Tuple(new_fields)
+
+    def compute_alignment(self, dtype):
+        dtype = TVMType(dtype)
+        align = (dtype.bits // 8) * dtype.lanes
+        # MAGIC CONSTANT FROM device_api.h
+        if align < 64:
+            align = 64
+
+        return expr.const(align, dtype="int64")
+
+    def compute_storage_in_relay(self, shape, dtype):
+        dtype = TVMType(dtype)
+        els = op.prod(shape)
+        num = expr.const(dtype.bits * dtype.lanes, self.compute_dtype) + expr.const(7, self.compute_dtype)
+        div = expr.const(8, self.compute_dtype)
+        return els * (num / div)
+
+    def compute_storage(self, tensor_type):
+        dtype = TVMType(tensor_type.dtype)
+        shape = [int(sh) for sh in tensor_type.shape]
+        size = 1
+        for i in range(len(shape)):
+            size *= shape[i]
+        size *= (dtype.bits * dtype.lanes + 7) // 8
+        return expr.const(size, dtype=self.compute_dtype)
+
+    def make_static_allocation(self, scope, tensor_type, i):
+        shape = [int(sh) for sh in tensor_type.shape]
+        if not len(shape):
+            shape = expr.const(np.array([]).astype(self.compute_dtype), dtype=self.compute_dtype)
+        else:
+            shape = expr.const(np.array(shape), dtype=self.compute_dtype)
+        size = self.compute_storage(tensor_type)
+        alignment = self.compute_alignment(tensor_type.dtype)
+        dtype = tensor_type.dtype
+        sto = scope.let(f"storage_{i}", self.alloc_storage(size, alignment, dtype))
+        tensor = self.alloc_tensor(sto, shape, dtype)
+        return scope.let(f"tensor_{i}", tensor)
+
+    def visit_let(self, let):
+        scope = ScopeBuilder()
+
+        self.scopes.append(scope)
+        while isinstance(let, expr.Let):
+            new_val = self.visit(let.value)
+            scope.let(let.var, new_val)
+            let = let.body
+
+        new_body = self.visit(let)
+        scope.ret(new_body)
+        self.scopes.pop()
+
+        return scope.get()
+
+    def visit_call(self, call):
+        if is_primitive(call):
+            # Because we are in ANF we do not need to visit the arguments.
+            scope = self.current_scope()
+            new_args = [self.visit(arg) for arg in call.args]
+            ins = expr.Tuple(new_args)
+            ret_type = call.checked_type
+
+            if ret_type.is_dynamic():
+                assert isinstance(ret_type, ty.TensorType)
+                shape_func_ins = []
+                engine = compile_engine.get()
+                cfunc = engine.lower_shape_func(call.op, self.target_host)
+                param_type = int(cfunc.shape_func_param_states[0])
+                for state in cfunc.shape_func_param_states:
+                    assert param_type == int(state)
+
+                for i, arg in enumerate(new_args):
+                    # Pass Shapes
+                    if param_type == 2:
```

Comment:
This part is incorrect. The size of param states should be same as the number of args. Each arg has its own corresponding shape func state: 0 - no data or shape required, 1 - only data required, 2 - only shape required, 3 - both required. 

Here you only use the first param state and apply to all args. Instead, you should zip shape_func_param_states and new_args.

Reply:
It looks like in the code there is a single flag for being shape dependent when you register the shape function, it simplifies everything if the arguments are all inputs or all shapes, do we really need a mixed mode?

Reply:
See the above code which asserts they are all equal

Reply:
I see. Yes, I agree it is more simple if they all share the same flags. I'm not sure whether we will have a mixed mode or not. If we fuse ops that have data independent shape func with ops that have data dependent shape func, it then becomes possible that the shape func for this fused op has a mixed mode.